kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,798 @@
+  local ipc_source = "dns_client_mlcache#" .. ipc_counter
+  local ipc = {
+    register_listeners = function(events)
+      -- The DNS client library will be required in globalpatches before Kong
+      -- initializes worker_events.
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local cwid = ngx.worker.id()
+      for _, ev in pairs(events) do
+        local handler = function(data, event, source, wid)
+          if cwid ~= wid then
+            ev.handler(data)
+          end
+        end
+
+        kong.worker_events.register(handler, ipc_source, ev.channel)
+      end
+    end,
+
+    broadcast = function(channel, data)
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local ok, err = kong.worker_events.post(ipc_source, channel, data)
+      if not ok then
+        log(ERR, "failed to post event '", ipc_source, "', '", channel, "': ", err)
+      end
+    end,
+  }
+
+  local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+    ipc             = ipc,
+    neg_ttl         = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    lru_size        = opts.cache_size or 10000,
+    resty_lock_opts = resty_lock_opts,
+  })
+
+  if not cache then
+    return nil, "could not create mlcache: " .. err
+  end
+
+  if opts.cache_purge then
+    cache:purge(true)
+  end
+
+  -- parse order
+  local search_types = {}
+  local order = opts.order or DEFAULT_ORDER
+  local preferred_ip_type
+  for _, typstr in ipairs(order) do
+    local qtype = valid_type_names[typstr:upper()]
+    if not qtype then
+      return nil, "Invalid dns record type in order array: " .. typstr
+    end
+
+    table_insert(search_types, qtype)
+
+    if (qtype == TYPE_A or qtype == TYPE_AAAA) and not preferred_ip_type then
+      preferred_ip_type = qtype
+    end
+  end
+  preferred_ip_type = preferred_ip_type or TYPE_A
+
+  if #search_types == 0 then
+    return nil, "Invalid order array: empty record types"
+  end
+
+  -- parse hosts
+  local hosts = init_hosts(cache, opts.hosts, preferred_ip_type)
+
+  return setmetatable({
+    cache         = cache,
+    stats         = {},
+    hosts         = hosts,
+    r_opts        = r_opts,
+    resolv        = opts._resolv or resolv,
+    valid_ttl     = opts.valid_ttl,
+    error_ttl     = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl     = opts.stale_ttl or DEFAULT_STALE_TTL,
+    empty_ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    search_types  = search_types,
+  }, mt)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = errcode == NAME_ERROR_EC and self.empty_ttl or self.error_ttl
+    -- compatible with balancer, which needs this field
+    answers.expire = now() + answers.ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = self.valid_ttl or 0xffffffff  -- 0xffffffff for maximum TTL value
```

Comment:
> Or do we want to validate?

For Kong, it's okay; the Kong configuration loader will perform the necessary checks. However, there's a possibility it might receive an incorrect `valid_ttl` from the PDK or custom plugins. Nevertheless, I believe it's currently sufficient for them. Third-party developers will encounter runtime errors and debug them to resolve any issues.

And accept ur suggested change.